eyeball-im-util/src/vector/sort.rs

```diff
@@ -0,0 +1,613 @@
+                                    new_values.pop_front().expect("`new_values` cannot be empty");
+
+                                buffered_vector.insert(index, (unsorted_index, new_value.clone()));
+                                result.push(
+                                    // At the beginning? Let's emit a `VectorDiff::PushFront`.
+                                    if index == 0 {
+                                        VectorDiff::PushFront { value: new_value }
+                                    }
+                                    // Somewhere in the middle? Let's emit a `VectorDiff::Insert`.
+                                    else {
+                                        VectorDiff::Insert { index, value: new_value }
+                                    },
+                                );
+
+                                // Finally, let's update the `last_insertion_index`.
+                                last_insertion_index = index;
+                            }
+                            None => {
+                                // `new_value` isn't consumed. Let's break the loop and emit a
+                                // `VectorDiff::Append` just after.
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                // Some values have not been inserted. Based on our algorithm, it means they
+                // must be appended.
+                if new_values.is_empty().not() {
+                    buffered_vector.append(new_values.clone());
+                    result.push(VectorDiff::Append {
+                        values: new_values.into_iter().map(|(_, value)| value).collect(),
+                    });
+                }
+            }
+        }
+        VectorDiff::Clear => {
+            // Nothing to do but clear.
+            buffered_vector.clear();
+            result.push(VectorDiff::Clear);
+        }
+        VectorDiff::PushFront { value: new_value } => {
+            // The unsorted index is inevitably 0, because we push a new item at the front
+            // of the vector.
+            let unsorted_index = 0;
+
+            // Shift all unsorted indices to the right.
+            // Also, find where to insert the `new_value`.
+            match buffered_vector
+                .iter_mut()
+                .enumerate()
+                .fold(None, |mut position, (index, (unsorted_index, value))| {
+                    if position.is_none() && compare(value, &new_value).is_ge() {
+                        position = Some(index);
+                    }
+
+                    *unsorted_index += 1;
+
+                    position
+                })
+                // If `buffered_vector` is empty, we want to emit a `VectorDiff::PushFront`. Let's
+                // map `position = None` to `position = Some(0)`.
+                .or_else(|| buffered_vector.is_empty().then_some(0))
+            {
```

Comment:
Could you extract the match scruntinee expression into a local binding for readability, please? Could be named something like `sorted_idx`.

Same for the `match <non-trivial expression>` expressions below.